torch/distributed/_functional_collectives.py

```diff
@@ -76,6 +76,221 @@ def is_torchdynamo_compiling():
 As a wise man said once: Don't cross the streams (https://www.youtube.com/watch?v=wyKQe_i9yyo)
 
 """
+
+"""
+Functional collectives can accept any of these types to describe the ranks participating in collectives.
+
+The different types will be desugared to a canonical format
+"""
+RANK_TYPES = Union[List[int], List[List[int]], dist.ProcessGroup, "dist._tensor.DeviceMesh", Tuple["dist._tensor.DeviceMesh", int]]
+
+
+"""
+User facing APIs for functional collectives
+-------------------------------------------
+
+These apis are called by usercode and expected to work both in eager execution and compilation,
+but there are significant differences to how the two modes are implemented underneath.
+
+Eager execution is 'optimized' using a tensor subclass that schedules the synchronization (via wait_tensor() op)
```

Comment:
nit: we should also probably remove or merge the previous comments L33-36 to this one?